gateway/handler.go

```diff
@@ -61,104 +90,156 @@ func Handler(gwapi lapi.Gateway, api lapi.FullNode, rateLimit int64, connPerMinu
 	m.Handle("/health/readyz", node.NewReadyHandler(api))
 	m.PathPrefix("/").Handler(http.DefaultServeMux)
 
-	/*ah := &auth.Handler{
-		Verify: nodeApi.AuthVerify,
-		Next:   mux.ServeHTTP,
-	}*/
-
-	rlh := NewRateLimiterHandler(m, rateLimit)
-	clh := NewConnectionRateLimiterHandler(rlh, connPerMinute)
-	return clh, nil
+	handler := &statefulCallHandler{m}
+	if perConnectionAPIRateLimit > 0 && perHostConnectionsPerMinute > 0 {
+		return NewRateLimitHandler(
+			handler,
+			perConnectionAPIRateLimit,
+			perHostConnectionsPerMinute,
+			connectionLimiterCleanupInterval,
+		), nil
+	}
+	return handler, nil
 }
 
-func NewRateLimiterHandler(handler http.Handler, rateLimit int64) *RateLimiterHandler {
-	limiter := limiterFromRateLimit(rateLimit)
-
-	return &RateLimiterHandler{
-		handler: handler,
-		limiter: limiter,
-	}
+type statefulCallHandler struct {
+	next http.Handler
 }
 
-// RateLimiterHandler adds a rate limiter to the request context for per-connection rate limiting
-type RateLimiterHandler struct {
-	handler http.Handler
-	limiter *rate.Limiter
+func (h statefulCallHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	tracker := newStatefulCallTracker()
+	defer func() {
+		go tracker.cleanup()
+	}()
+	r = r.WithContext(context.WithValue(r.Context(), statefulCallTrackerKey, tracker))
+	h.next.ServeHTTP(w, r)
 }
 
-func (h RateLimiterHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
-	r = r.WithContext(context.WithValue(r.Context(), perConnLimiterKey, h.limiter))
+func (h statefulCallHandler) Shutdown(ctx context.Context) error {
+	return shutdown(ctx, h.next)
+}
 
-	// also add a filter tracker to the context
-	r = r.WithContext(context.WithValue(r.Context(), statefulCallTrackerKey, newStatefulCallTracker()))
+type hostLimiter struct {
+	limiter    *rate.Limiter
+	lastAccess time.Time
+}
 
-	h.handler.ServeHTTP(w, r)
+type RateLimitHandler struct {
+	cancelFunc                  context.CancelFunc
+	mu                          sync.Mutex
+	limiters                    map[string]*hostLimiter
```

Comment:
I'm not actually sure anyone uses this! You'd be better to deploy this with nginx in front of it and use that for standard rate limiting and only using lotus-gateway to deal with the complexities of lotus calls, like the API rate limiting—but even that, you probably should gate users with API keys and handle them that way.

Basically: I'm not sure any of this rate limiting stuff is used in practice anywhere or is all that useful, I'm fixing what's there in case someone is using it!